wolfcrypt/src/port/maxim/maxq10xx.c

```diff
@@ -0,0 +1,3376 @@
+        *keycomplen = 66;
+        return MXQ_KEYPARAM_EC_P521R1;
+    case ECC_BRAINPOOLP256R1_OID:
+        *keycomplen = 32;
+        return MXQ_KEYPARAM_EC_BP256R1;
+    case ECC_BRAINPOOLP384R1_OID:
+        *keycomplen = 48;
+        return MXQ_KEYPARAM_EC_BP384R1;
+    case ECC_BRAINPOOLP512R1_OID:
+        *keycomplen = 64;
+        return MXQ_KEYPARAM_EC_BP512R1;
+    default:
+        return MXQ_UNKNOWN_CURVE;
+    }
+}
+
+static int mxq_get_sign_alg_from_sig_oid(word32 maxq_id) {
+
+    switch (maxq_id) {
+    case CTC_SHA256wECDSA:
+        return ALGO_ECDSA_SHA_256;
+    case CTC_SHA384wECDSA:
+        return ALGO_ECDSA_SHA_384;
+    case CTC_SHA512wECDSA:
+        return ALGO_ECDSA_SHA_512;
+    case CTC_SHA256wRSA:
+        return ALGO_RSASSAPSSPKCS1_V1_5_SHA256;
+    case CTC_SHA384wRSA:
+        return ALGO_RSASSAPSSPKCS1_V1_5_SHA384;
+    case CTC_SHA512wRSA:
+        return ALGO_RSASSAPSSPKCS1_V1_5_SHA512;
+    default:
+        return ALGO_INVALID;
+    }
+}
+#endif /* WOLFSSL_MAXQ108X */
+
+/* TODO: Get rid of publicKeyIndex. Pass in the information it is parsing.
+ *       Then Improve this to pass in the the parts of decoded cert it needs.
+ */
+static int maxq10xx_process_server_certificate(WOLFSSL* ssl,
+                                               DecodedCert* p_cert)
+{
+    (void)ssl;
+    mxq_keytype_id_t key_type = MXQ_KEYTYPE_ECC;
+    mxq_keyparam_id_t keyparam = MXQ_KEYPARAM_EC_P256R1;
+    mxq_length totalkeylen;
+    mxq_algo_id_t sign_algo = ALGO_ECDSA_SHA_256;
+    int pk_offset = p_cert->publicKeyIndex;
+    mxq_length keycomplen = 32;
+    int rc;
+    mxq_err_t mxq_rc;
+    mxq_u1 certdata[2048];
+    mxq_length certdatalen = sizeof(certdata);
+#if defined(WOLFSSL_MAXQ1065)
+
+    if (p_cert->signatureOID != CTC_SHA256wECDSA) {
+        WOLFSSL_MSG("MAXQ: signature algo not supported");
+        return NOT_COMPILED_IN;
+    }
+
+    if (p_cert->keyOID != ECDSAk) {
+        WOLFSSL_MSG("MAXQ: key algo not supported");
+        return NOT_COMPILED_IN;
+    }
+
+    if (p_cert->pkCurveOID != ECC_SECP256R1_OID) {
+        WOLFSSL_MSG("MAXQ: key curve not supported");
+        return NOT_COMPILED_IN;
+    }
+
+    totalkeylen = keycomplen * 2;
+
+#elif defined(WOLFSSL_MAXQ108X)
+
+    if (p_cert->keyOID == ECDSAk )
+    {
+        keyparam = maxq_curve_wolfssl_id2mxq_id(p_cert->pkCurveOID,
+                                                &keycomplen);
+        if (keyparam == MXQ_UNKNOWN_CURVE) {
+            WOLFSSL_MSG("MAXQ: key curve not supported");
+            return NOT_COMPILED_IN;
+        }
+        totalkeylen = keycomplen * 2;
+    }
+    else if (p_cert->keyOID == RSAk) {
+        pk_offset = calculate_modulus_offset(p_cert->source,
+                                             p_cert->publicKeyIndex+1);
+        keycomplen = ((p_cert->source[pk_offset-2] << 8) |
+                      p_cert->source[pk_offset-1]);
+
+        /* Is there a more elegant way for checking pub key??? */
```

Comment:
This is trying to find the exponent value. Code is terrible.

Reply:
Put a TODO at top of function to improve on this.